### PR TITLE
Refactor/repository layer

### PR DESCRIPTION
**Pull request by jurgenbelien**
# Changes

## Added

* Generic Repository class to add an abstraction to firebase
* Types to more easily interface with the repositories
* Generic Service for interfacing with Repository
* Generic RecordsService for interfacing with Repository of Records (items that have date and employeeId fields)

## Changed

* Services now use these abstractions

## How to test

* Visit pages that fetch different resources from firebase

# To Do
- [ ] Fix employeeService.getByEmail error
- [ ] Implement StandbyRecordsService
- [ ] See if the field unsetting in CustomersService is needed
- [ ] Clean up typing

### CODE REVIEW

File: services/Service.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    return this.repository.getByIds(ids);
+  }
+
+  add(resource: T) {
+    return this.repository.add(resource);
+  }
+
+  update(resource: DocumentWithId<T>) {
+    const {id, ...doc} = resource;
+    return this.repository.updateById(id, doc as unknown as T); // TODO: fix type
```
Comment (marioesuc) [inline services/Service.ts:30]: You're leaving this for another PR I assume?

File: services/Service.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import RepositoryManager, {DocumentData, DocumentWithId, Repository} from '~/repositories';
+
+export default class Service<T extends DocumentData = DocumentData> {
+  repositories: RepositoryManager;
```
Comment (marioesuc) [inline services/Service.ts:4]: I don't really get this, why is it necessary to pass all the repositories? I don't see it being used anywhere in the class, isn't it just enough to pass the concrete one?

File: services/customers-service.ts [hunk 3]
```diff
@@ -1,73 +1,44 @@
...
   }
 
-  async updateCustomer(customer: Customer) {
-    const newCustomer = {...customer} as any;
-    delete newCustomer.id;
+  updateCustomer(customer: DocumentWithId<Customer>) {
+    return this.update(customer);
+
+    // TODO: Figure out how to remove fields without having firestore complexity leak into here
+    //   const newCustomer = {...customer} as any;
```
Comment (marioesuc) [inline services/customers-service.ts:30]: Where is all this logic happening now if it's commented out?

File: services/employees-service.ts [hunk 4]
```diff
@@ -1,155 +1,110 @@
...
   }
 
-  private mapEmployeeFromDocument(doc: DocumentSnapshot): Omit<Employee, 'isAdmin'> {
-    return {
-      ...(doc.data() as Omit<Employee, 'isAdmin'>),
-      id: doc.id,
-    };
+  private async revokeAdmin({email}: Employee) {
+    const {admins = [], id} = (await this.repositories.adminList.getByIndex(0)) || {};
+    const shouldUpdate =
```
Comment (marioesuc) [inline services/employees-service.ts:99]: `makeAdmin` and `revokeAdmin` have identical implementations except for the `shouldUpdate` const and condition. We could make it one function and pass this as a callback via parameters that is called after fetching the `admins` array and pass it into the callback.

File: services/Service.ts (hunk 1 above)
Comment (jurgenbelien) [reply services/Service.ts:30]: Good catch! I should revisit this: the `type DocumentData = FBDocumentData & { id: never }; ` in Repository.ts should make this unnecessary!

File: services/customers-service.ts (hunk 3 above)
Comment (jurgenbelien) [reply services/customers-service.ts:30]: It isn't, I need to know if this is really necessary though.

File: services/employees-service.ts (hunk 4 above)
Comment (jurgenbelien) [reply services/employees-service.ts:99]: I can make another private method for the update, but I really want to make sure that revoke and make are separate functions for clarity's sake

File: services/Service.ts (hunk 2 above)
Comment (jurgenbelien) [reply services/Service.ts:4]: Some services call multiple repositories and need access to multiple repositories, for instance EmployeesService. That service needs both `RepositoryManager.prototype.employees` and `RepositoryManager.prototype.adminList`.

The `Service.prototype.repository` on the other hand is a convenience method to bind the primary repository, as is visible in the `RecordsService` subclass, so you don't have to repeat the name.

**Pull request closed without merging**